vars/pythonSetupPyPipeline.groovy

```diff
@@ -9,148 +9,310 @@
+                junit "build/test_results.xml"
+                cobertura coberturaReportFile: "build/coverage.xml"
+              }
+            }
+          } // Run Tests
         }
-      }
+      } // Validate
 
       stage("Package") {
+        agent {
+          docker {
+            image buildImage.imageName()
+            args pipelineParams.dockerRunArgs
+            reuseNode true
+          }
+        }
         steps {
           sh "python setup.py sdist"
           sh "twine check dist/*"
+
+          script {
+            // Gather the module name and version so that it can be used in the Docker deploy stage
+            moduleName = sh(script: "python setup.py --name", returnStdout: true).trim()
+            moduleVersion = sh(script: "python setup.py --version", returnStdout: true).trim()
+          }
         }
-      }
+      } // Package
 
       stage("Deploy") {
-        when {
-          beforeAgent true
-          allOf {
-            branch "master"
-            expression {
-              params.doRelease
+        parallel {
+          stage("Deploy PyPI") {
+            agent {
+              docker {
+                image buildImage.imageName()
+                args pipelineParams.dockerRunArgs
+                reuseNode true
+              }
             }
-          }
+            when {
+              beforeAgent true
+              allOf {
+                branch "master"
+                expression { params.doRelease }
+              }
+            }
+            steps {
+              withCredentials([usernamePassword(credentialsId: pipelineParams.pypiCredentialsId, usernameVariable: "USERNAME", passwordVariable: "PASSWORD")]) {
+                sh "twine upload --verbose -u $USERNAME -p $PASSWORD --repository-url ${pipelineParams.pypiRepo} dist/*"
+              }
+            }
+          } // Deploy PyPI
+
+          stage("Deploy Docker") {
+            when {
+              beforeAgent true
+              expression {
+                // On master builds deploy always if dockerDeploy == true.
+                // On branch/PR builds deploy only on-demand.
+                (BRANCH_NAME ==~ /(master)/ && pipelineParams.dockerDeploy) || params.doDockerSnapshot
+              }
+            }
+            steps {
+              script {
+                docker.withRegistry(pipelineParams.dockerRegistryUrl, pipelineParams.dockerRegistryCredentialsId) {
+                  def image = docker.image("${pipelineParams.dockerRepo}/${moduleName}:${moduleVersion}")
+                  def imageExists = true
+                  try {
+                    image.pull()
+                  } catch(Exception) {
+                    echo "Image '${image.imageName()}' already exists in registry"
+                    imageExists = false
+                  }
+
+                  // Snapshot images can be overwritten, whereas a release one shouldn't be.
+                  if (isSnapshot(moduleVersion) || !imageExists) {
```

Comment:
Don't you have here a diaper pattern? In case for example a release version is already present in the docker registry for whatever reason you will silently continue, while effectively your release procedure isn't completed as expected.

Reply:
You are right, better to stop the whole pipeline if that situation occurs. Updated this piece in the latest commit